lib/appsignal/helpers/instrumentation.rb

```diff
@@ -5,6 +5,144 @@ module Helpers
+      #       :namespace => "my_other_namespace",
+      #       :action => "MyOtherClass#my_other_method"
+      #     ) do
+      #       # Some code
+      #
+      #       # The reported namespace will be "my_namespace"
+      #       # The reported action will be "MyClass#my_method"
+      #     end
+      #   end
+      #
+      # @param namespace [String/Symbol] The namespace to set on the new
+      #   transaction.
+      #   Defaults to the 'web' namespace.
+      #   This will update the already active transaction's namespace if
+      #   {.monitor} is called when another transaction is already active.
+      # @param action [String/Symbol] The action name for the transaction.
+      #   This will update the already active transaction's action if
+      #   {.monitor} is called when another transaction is already active.
+      # @yield The block to monitor.
+      # @raise [Exception] Any exception that occurs within the given block is
+      #   re-raised by this method.
+      # @return [Object] The value of the given block is returned.
+      # @since 3.11.0
+      def monitor(
+        namespace: nil,
+        action: nil
+      )
+        return yield unless active?
+
+        has_parent_transaction = Appsignal::Transaction.current?
+        transaction =
+          if has_parent_transaction
+            Appsignal::Transaction.current
+          else
+            Appsignal::Transaction.create(
+              SecureRandom.uuid,
+              namespace || Appsignal::Transaction::HTTP_REQUEST,
+              Appsignal::Transaction::GenericRequest.new({})
+            )
+          end
+
+        begin
+          yield if block_given?
+        rescue Exception => error # rubocop:disable Lint/RescueException
+          transaction.set_error(error)
+          raise error
+        ensure
+          if has_parent_transaction
+            if namespace
+              callers = caller
+              Appsignal::Utils::StdoutAndLoggerMessage.warning \
+                "A parent transaction is active around this 'Appsignal.monitor' call. " \
+                  "The namespace is not updated for the parent transaction." \
+                  "If you want to update the namespace for this parent transaction, " \
+                  "call 'Appsignal.set_namespace' from within the 'Appsignal.monitor' block. " \
+                  "Update the 'Appsignal.monitor' call in: #{callers.first}"
+            end
+            if action
+              callers = caller
+              Appsignal::Utils::StdoutAndLoggerMessage.warning \
+                "A parent transaction is active around this 'Appsignal.monitor' call. " \
+                  "The action is not updated for the parent transaction." \
+                  "If you want to update the action for this parent transaction, " \
+                  "call 'Appsignal.set_action' from within the 'Appsignal.monitor' block. " \
+                  "Update the 'Appsignal.monitor' call in: #{callers.first}"
```

Comment:
I wonder if it would be possible to summarise this as a single "calling `.monitor` inside `.monitor` has no effect" message?

I also don't think these need to be warnings -- I think it's okay that calling `.monitor` inside `.monitor` is a no-op. I would suggest `debug`, maybe.